Compile fail (undefined reference to `pthread_create')

Hi
There is an error message When I build this project, The message as below: 

[ 16%] Linking CXX executable fibonacci
CMakeFiles/fibonacci.dir/examples/fibonacci.cpp.o: In function `std::thread::thread<lars::Generator<long>::begin() const::{lambda()#1}>(lars::Generator<long>::begin() const::{lambda()#1}&&)':
fibonacci.cpp:(.text._ZNSt6threadC2IZNK4lars9GeneratorIlE5beginEvEUlvE_JEEEOT_DpOT0_[_ZNSt6threadC5IZNK4lars9GeneratorIlE5beginEvEUlvE_JEEEOT_DpOT0_]+0x2f): undefined reference to `pthread_create'
collect2: error: ld returned 1 exit status
CMakeFiles/fibonacci.dir/build.make:94: recipe for target 'fibonacci' failed
make[2]: *** [fibonacci] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/fibonacci.dir/all' failed
make[1]: *** [CMakeFiles/fibonacci.dir/all] Error 2
Makefile:94: recipe for target 'all' failed
make: *** [all] Error 2

My solution:
Link pthread to target(example/test)

target_link_libraries( ${examplename} pthread)
target_link_libraries( ${testname} pthread)

--- a/.travis.yml
+++ b/.travis.yml
@@ -46,7 +46,10 @@ install:
   - sudo cmake --build build --target install
 
 script:
+  # tests
   - cmake -Htests -Bbuild/test  
   - cmake --build build/test
   - CTEST_OUTPUT_ON_FAILURE=1 cmake --build build/test --target test
- 
+ # build examples
+  - cmake -Hexamples -Bbuild/examples
+  - cmake --build build/examples

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,0 +1,52 @@
+
+language: cpp
+sudo: require
+
+common_sources: &all_sources
+  - ubuntu-toolchain-r-test
+  - llvm-toolchain-trusty
+  - llvm-toolchain-trusty-6.0
+
+matrix:
+  include:
+    - os: linux
+      compiler: gcc
+      addons: &gcc8
+        apt:
+          sources: *all_sources
+          packages: 
+            - g++-8
+      env:
+        - MATRIX_EVAL="export CC=gcc-8; export CXX=g++-8;"
+
+    - os: linux
+      compiler: clang
+      addons:
+          apt:
+              sources: *all_sources
+              packages:
+                - g++-8
+                - clang-6.0
+      env:
+        - MATRIX_EVAL="export CC=clang-6.0; export CXX=clang++-6.0;"
+
+before_install:
+  # Update compilers
+  - eval "${MATRIX_EVAL}"
+  - echo "CC=$CC CXX=$CXX"
+  # Install a supported cmake version (>= 3.14)
+  - wget -O cmake.sh https://cmake.org/files/v3.14/cmake-3.14.0-Linux-x86_64.sh 
+  - sudo sh cmake.sh --skip-license --exclude-subdir --prefix=/usr/local
+  - export PATH=/usr/local/bin:$PATH
+  - cmake --version
+
+install:
+  - cmake -H. -Bbuild
+  - cmake --build build
+  - sudo cmake --build build --target install
+
+script:
+  - cmake -Htests -Bbuild/test  
+  - cmake --build build/test
+  - CTEST_OUTPUT_ON_FAILURE=1 cmake --build build/test --target test
+ 

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -8,7 +8,8 @@ project(LarsGenerator
 )
 # ---- Configuration variables ----
 
-option(LarsGenerator_ENABLE_TESTS "Enable tests" OFF)
+option(LarsGenerator_ENABLE_TESTS "Enable LarsGenerator tests" OFF)
+option(LarsGenerator_BUILD_EXAMPLES "Build LarsGenerator examples" OFF)
 
 # ---- Include guards ----
 
@@ -83,9 +84,13 @@ install(
   DESTINATION include
 )
 
-# ---- Test ----
+# ---- Tests and examples ----
 
-if(${LarsGenerator_ENABLE_TESTS})
+if (${LarsGenerator_BUILD_EXAMPLES})
+  add_subdirectory(examples)  
+endif()
+
+if (${LarsGenerator_ENABLE_TESTS})
   ENABLE_TESTING()  
   add_subdirectory(tests)  
 endif()

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,45 +1,92 @@
-cmake_minimum_required(VERSION 3.1)
+cmake_minimum_required(VERSION 3.5 FATAL_ERROR)
 
-get_filename_component(PROJECT_NAME ${CMAKE_CURRENT_SOURCE_DIR} NAME)
-SET(PROJECT_NAME Lars${PROJECT_NAME})
-project(${PROJECT_NAME})
+# ---- Project ----
+
+project(LarsGenerator 
+  VERSION 0.1
+  LANGUAGES CXX
+)
+# ---- Configuration variables ----
+
+option(LarsGenerator_ENABLE_TESTS "Enable tests" OFF)
+
+# ---- Include guards ----
+
+if(TARGET LarsGenerator)
+  return()
+endif()
+
+if(${PROJECT_SOURCE_DIR} STREQUAL ${PROJECT_BINARY_DIR})
+    message(FATAL_ERROR "In-source builds not allowed. Please make a new directory (called a build directory) and run CMake from there.")
+endif()
+
+# ---- Header target ----
 
 FILE(GLOB headers "${CMAKE_CURRENT_SOURCE_DIR}/include/lars/*.h")
-get_directory_property(has_parent PARENT_DIRECTORY)
+add_library(LarsGenerator-headers EXCLUDE_FROM_ALL ${headers})
+SET_TARGET_PROPERTIES(LarsGenerator-headers PROPERTIES LINKER_LANGUAGE CXX)
 
-IF(LARS_LIBRARY)
-  SET(lars_headers ${lars_headers} ${headers} PARENT_SCOPE)
-  SET(lars_include_dirs ${lars_include_dirs} "${CMAKE_CURRENT_SOURCE_DIR}/include" PARENT_SCOPE)
-ELSEIF(has_parent)
-  SET(include_dirs ${CMAKE_CURRENT_SOURCE_DIR}/include ${CMAKE_CURRENT_SOURCE_DIR}/bundled/include)
-  SET(${PROJECT_NAME}_INCLUDE_DIRS ${include_dirs} CACHE INTERNAL "")
-ELSE()
-  SET(LARS_BUILD_EXAMPLES TRUE)
-  SET(LARS_BUILD_TESTS TRUE)
-ENDIF()
+# ---- Create library ----
 
-IF(LARS_BUILD_EXAMPLES)
-  SET(include_dirs ${CMAKE_CURRENT_SOURCE_DIR}/include ${CMAKE_CURRENT_SOURCE_DIR}/bundled/include)
-  include_directories(${include_dirs})
-  set(CMAKE_CXX_STANDARD 11)
-  file( GLOB example_sources ${CMAKE_CURRENT_SOURCE_DIR}/examples/*.cpp )
-  foreach( examplesourcefile ${example_sources} )
-    get_filename_component(filename ${examplesourcefile} NAME)
-    string( REPLACE ".cpp" "" examplename ${filename} )
-    add_executable( ${examplename} ${examplesourcefile} ${headers})
-  endforeach()
-ENDIF()
+add_library(LarsGenerator INTERFACE)
+target_link_libraries(LarsGenerator INTERFACE pthread)
 
-IF(LARS_BUILD_TESTS)
-  enable_testing()
-  SET(include_dirs ${CMAKE_CURRENT_SOURCE_DIR}/include ${CMAKE_CURRENT_SOURCE_DIR}/bundled/include)
-  include_directories(${include_dirs})
-  set(CMAKE_CXX_STANDARD 11)
-  file( GLOB test_sources ${CMAKE_CURRENT_SOURCE_DIR}/tests/*.cpp )
-  foreach( testsourcefile ${test_sources} )
-    get_filename_component(filename ${testsourcefile} NAME)
-    string( REPLACE ".cpp" "" testname ${filename} )
-    add_executable(${testname} ${testsourcefile} ${headers})
-    add_test(NAME ${testname} COMMAND ${testname})
-  endforeach()
-ENDIF()
+target_include_directories(LarsGenerator
+  INTERFACE 
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
+    $<INSTALL_INTERFACE:include>
+)
+
+include(CMakePackageConfigHelpers)
+
+write_basic_package_version_file(
+  "${PROJECT_BINARY_DIR}/LarsGeneratorConfigVersion.cmake"
+  VERSION ${PROJECT_VERSION}
+  COMPATIBILITY AnyNewerVersion
+)
+
+# ---- Install ----
+
+install(
+  TARGETS LarsGenerator
+  EXPORT LarsGeneratorTargets
+  LIBRARY DESTINATION lib COMPONENT Runtime
+  ARCHIVE DESTINATION lib COMPONENT Development
+  RUNTIME DESTINATION bin COMPONENT Runtime
+  PUBLIC_HEADER DESTINATION include COMPONENT Development
+  BUNDLE DESTINATION bin COMPONENT Runtime
+)
+
+include(CMakePackageConfigHelpers)
+
+configure_package_config_file(
+  "${PROJECT_SOURCE_DIR}/cmake/LarsGeneratorConfig.cmake.in"
+  "${PROJECT_BINARY_DIR}/LarsGeneratorConfig.cmake"
+  INSTALL_DESTINATION lib/cmake/LarsGenerator
+)
+
+install(
+  EXPORT LarsGeneratorTargets 
+  DESTINATION lib/cmake/LarsGenerator
+)
+
+install(
+  FILES 
+    "${PROJECT_BINARY_DIR}/LarsGeneratorConfigVersion.cmake"
+    "${PROJECT_BINARY_DIR}/LarsGeneratorConfig.cmake"
+  DESTINATION 
+    lib/cmake/LarsGenerator
+)
+
+install(
+  DIRECTORY ${PROJECT_SOURCE_DIR}/include/
+  DESTINATION include
+)
+
+# ---- Test ----
+
+if(${LarsGenerator_ENABLE_TESTS})
+  ENABLE_TESTING()  
+  add_subdirectory(tests)  
+endif()
+

--- a/README.md
+++ b/README.md
@@ -1,14 +1,21 @@
+[![Build Status](https://travis-ci.com/TheLartians/Generator.svg?branch=master)](https://travis-ci.com/TheLartians/Generator)
+
 # Generator
 A python-style generator class emulating yield functionality for simple and memory-efficient iterator creation.
 
-## Clone and build examples
+# Example
 
 ```
-git clone git@github.com:TheLartians/Generator.git
-cd Generator
-mkdir build
-cd build
-cmake ..
-make
-```
+#include <lars/generator.h>
+#include <iostream>
 
+lars::Generator<int> range(int max){
+  return lars::Generator<int>([=](lars::Yield<int> &yield){
+    for(int i = 0;i<max;++i) yield(i);
+  });
+}
+
+int main(){
+  for(int i:range(10)) std::cout << i << std::endl;
+}
+```

--- a/cmake/LarsGeneratorConfig.cmake.in
+++ b/cmake/LarsGeneratorConfig.cmake.in
@@ -1,0 +1,5 @@
+@PACKAGE_INIT@
+
+include("${CMAKE_CURRENT_LIST_DIR}/LarsGeneratorTargets.cmake")
+check_required_components("@PROJECT_NAME@")
+

--- a/examples/CMakeLists.txt
+++ b/examples/CMakeLists.txt
@@ -1,0 +1,30 @@
+cmake_minimum_required(VERSION 3.5 FATAL_ERROR)
+
+# ---- Project ----
+
+project(Examples CXX)
+
+# ---- CXX Flags ----
+
+if(NOT CMAKE_CXX_STANDARD GREATER 17)
+  set(CMAKE_CXX_STANDARD 17)
+endif()
+
+# ---- Requires ----
+
+if(CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
+  find_package(LarsGenerator 0.1 REQUIRED)
+endif()
+
+# ---- Create binaries ----
+
+file(GLOB example_sources ${CMAKE_CURRENT_SOURCE_DIR}/*.cpp)
+
+foreach( example_source_file ${example_sources} )
+  get_filename_component(filename ${example_source_file} NAME)
+  string(REPLACE ".cpp" "" example_name ${filename} )
+  add_executable(${example_name} ${example_source_file})
+  target_link_libraries(${example_name} LarsGenerator)  
+  set_target_properties(${example_name} PROPERTIES CXX_STANDARD 17)
+endforeach()
+

--- a/examples/fibonacci.cpp
+++ b/examples/fibonacci.cpp
@@ -5,7 +5,6 @@
 using namespace lars;
 using namespace std;
 
-
 int main(){
 
   auto fibonacci_numbers = Generator<long>([=](Yield<long> &yield){
@@ -21,9 +20,8 @@ int main(){
   });
   
   for(auto i:fibonacci_numbers){
-     if(i > 1000000) break;
+     if(i > 50) break;
      std::cout << i << std::endl;
   }
   
 }
-

--- a/examples/fibonacci.cpp
+++ b/examples/fibonacci.cpp
@@ -20,8 +20,8 @@ int main(){
   });
   
   for(auto i:fibonacci_numbers){
-     if(i > 50) break;
      std::cout << i << std::endl;
+     if(i > 10000) break;
   }
   
 }

--- a/examples/range.cpp
+++ b/examples/range.cpp
@@ -2,16 +2,13 @@
 #include <lars/generator.h>
 #include <iostream>
 
-using namespace lars;
-using namespace std;
-
-Generator<int> range(int max){
-  return Generator<int>([=](Yield<int> &yield){
+lars::Generator<int> range(int max){
+  return lars::Generator<int>([=](lars::Yield<int> &yield){
     for(int i = 0;i<max;++i) yield(i);
   });
 }
 
 int main(){
-  for(int i:range(10)) cout << i << endl;
+  for(int i:range(10)) std::cout << i << std::endl;
 }
 

--- a/include/lars/generator.h
+++ b/include/lars/generator.h
@@ -60,6 +60,7 @@ namespace lars{
     Yield(){ }
     
   public:
+  
     void operator()(const T &value){
       std::unique_lock<std::mutex> lock(mutex);
       while(state == WAITING) ready_wait.wait(lock);
@@ -89,19 +90,26 @@ namespace lars{
       std::unique_ptr<Data> data;
       
     public:
-      const T &operator*(){ return data->current_value; }
-      const T *operator->(){ return &data->current_value; }
+
+      T &operator*() const {
+        return data->current_value;
+      }
       
-      void operator++(){
+      T *operator->() const {
+        return &data->current_value;
+      }
+      
+      const_iterator &operator++(){
         try{
           data->current_value = data->yield.get_future().get();
         }
         catch(typename Yield<T>::FinishedException){
           data.reset();
         }
+        return *this;
       }
       
-      bool operator!=(const const_iterator &other){ return data != other.data; }
+      bool operator!=(const const_iterator &other)const{ return data != other.data; }
     };
     
     const_iterator begin()const{

--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -1,0 +1,33 @@
+cmake_minimum_required(VERSION 3.14 FATAL_ERROR)
+include(FetchContent)
+
+# ---- Project ----
+
+project(LarsGeneratorTests CXX)
+
+# ---- Requires ----
+
+if(CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
+  find_package(LarsGenerator REQUIRED)
+endif()
+
+FetchContent_Declare(
+  Catch2
+  GIT_REPOSITORY https://github.com/catchorg/Catch2.git
+  GIT_TAG        v2.5.0
+)
+
+FetchContent_MakeAvailable(Catch2)
+
+# ---- Create binary ----
+
+file(GLOB tests_sources ${CMAKE_CURRENT_SOURCE_DIR}/*.cpp)
+add_executable(lars-generator-tests ${tests_sources})
+target_link_libraries(lars-generator-tests LarsGenerator Catch2)
+set_target_properties(lars-generator-tests PROPERTIES CXX_STANDARD 17 COMPILE_FLAGS "${CMAKE_CXX_FLAGS} -Wall -pedantic -Wextra -Werror")
+
+# ---- Add tests ----
+
+ENABLE_TESTING() 
+ADD_TEST(lars-generator-tests lars-generator-tests)
+

--- a/tests/generator.cpp
+++ b/tests/generator.cpp
@@ -1,0 +1,52 @@
+#include <catch2/catch.hpp>
+#include <iterator>
+
+#include <lars/generator.h>
+
+TEST_CASE("Range Generator") {
+
+  using namespace lars;
+
+  auto range = [](auto max){
+    using T = decltype(max);
+    return Generator<T>([=](Yield<T> &yield){
+      for(T i = 0;i<max;++i) yield(i);
+    });
+  };
+
+  std::vector<int> v;
+  for (auto i: range(10)) {
+    v.push_back(i);
+  }
+
+  REQUIRE(v.size() == 10);
+  
+  for (int i=0; i<10; ++i) {
+    REQUIRE(v[i] == i);
+  }
+
+}
+
+TEST_CASE("Fibonacci Generator") {
+  using namespace lars;
+  using Number = unsigned long long;
+
+  auto fibonacci_numbers = Generator<Number>([=](Yield<Number> &yield){
+    Number a = 0, b = 1;
+    yield(a);
+    yield(b);
+    while(true){
+      Number t = a + b;
+      yield(t);
+      a = b;
+      b = t;
+    }
+  });
+
+  auto it = fibonacci_numbers.begin();
+  
+  std::advance(it, 55);
+  REQUIRE((*it) == 139583862445ll);
+}
+
+

--- a/tests/main.cpp
+++ b/tests/main.cpp
@@ -1,0 +1,3 @@
+#define CATCH_CONFIG_MAIN
+
+#include <catch2/catch.hpp>

